### PR TITLE
Add experimental feature support for enabling Windows Feature dependencies

### PR DESCRIPTION
**Pull request by ryfu-msft**
Related to: #556 

Description:
This PR adds experimental feature support for enabling Windows Features prior to installation. There are known packages such as Wix toolkit that require certain optional Windows Features to be enabled before installation can run to completion. 

Changes:
- Added new `windowsFeature` experimental feature to the settings schema.
- Enabling windows feature is done by loading the DismApi.dll and calling those dll functions. 

Implementation flow:
- Checks if experimental feature is enabled
- Checks if admin mode is enabled in order to support calling DismApis
- Retrieves the WindowsFeatures dependencies and verifies that they all exist, if they do not, winget exits with a friendly error message. A user can add the `--ignore-missing-dependencies` flag to bypass this check and continue with the installation.
- Each valid Windows Feature dependency is enabled if not already enabled. If any of the dependencies fail to be enabled, an error message with the corresponding HRESULT is displayed to the user. The user can add `--force` to bypass this failure and continue with installation. Often times, a reboot will be required to fully enable a feature, in this scenario we warn the user but continue with the installation as it is possible that the installation may still succeed.

Tests:
- Added unit tests for verifying the behavior of the CLI output based on the exit codes from enabling/disabling Windows Features using TestHooks
- Added tests enabling `netfx3` and disabling `netfx3` and verifying the output is successful.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/3005)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 1]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+    void EnableWindowsFeaturesDependencies(Execution::Context& context)
+    {
+        if (!Settings::ExperimentalFeature::IsEnabled(Settings::ExperimentalFeature::Feature::WindowsFeature))
+        {
+            return;
+        }
+
+        const auto& rootDependencies = context.Get<Execution::Data::Installer>()->Dependencies;
+
+        if (rootDependencies.Empty())
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:147]: Is this already handled by `HasAnyOf` ? If not, should it be?

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 2]
```diff
@@ -1677,4 +1677,36 @@ Please specify one of them using the --source option to proceed.</value>
...
+    <comment>"Windows Feature(s)" is the name of the options Windows features setting.
+{Locked="--force"}
+{Locked="--ignore-missing-dependencies"}</comment>
+  </data>
+  <data name="IgnoreMissingDependenciesArgumentDescription" xml:space="preserve">
+    <value>Ignores missing Windows Feature dependencies</value>
+    <comment>"Windows Feature" is the name of the options Windows features setting.</comment>
+  </data>
+  <data name="EnablingWindowsFeatures" xml:space="preserve">
+    <value>Enabling Windows Feature dependencies...</value>
```
Comment (Trenly) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1697]: As a user, I would like to know specifically which features are being enabled
```
    <value>Enabling Windows Feature {0}.</value>
```

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 3]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+        {
+            return;
+        }
+
+        if (rootDependencies.HasAnyOf(DependencyType::WindowsFeature))
+        {
+            context << Workflow::EnsureRunningAsAdmin;
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:154]: Installer dependencies are known before this workflow is ever executed; Specific dependency types are also checked for in the `ReportDependencies` workflow.

I'm wondering, now that multiple dependency types will be supported, if it makes more sense to set an execution flag for "Contains Package Dependencies" or "Contains Windows Features Dependencies" in the ReportDependencies flow.

With the flags, then would be possible to have a separate workflow around `CheckSupportForWindowsFeatures` that checks the flag ensures the experimental feature is enabled and user is running as admin, to avoid creating copies of the dependencies data in memory unnecessarily

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 4]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+                {
+                    AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_INSTALL_MISSING_DEPENDENCY);
+                }
+            }
+
+            HRESULT hr = S_OK;
+            bool continueOnFailure = context.Args.Contains(Execution::Args::Type::Force);
+            bool rebootRequired = false;
+
+            context.Reporter.Info() << Resource::String::EnablingWindowsFeatures << std::endl;
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:203]: As a user, I would prefer this be shown per-feature as they are enabled; I don't need to be informed that features dependencies are being enabled in the case that all the features are already enabled

File: doc/Settings.md [hunk 5]
```diff
@@ -263,3 +263,14 @@ You can enable the feature as shown below.
...
 ```
+
+### windowsFeature
+
+This feature enables the ability to enable Windows Feature dependencies during installation.
+You can enable the feature as shown below.
+
+```json
+   "experimentalFeatures": {
+       "windowsFeature": true
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply doc/Settings.md:274]: Features are a code construct whose state is driven by settings.  Settings can theoretically be arbitrarily complex, but I don't think we have any that are not simple types currently.  The settings code could even handle both bool and object values for `dependencies` if we were inclined.

[the remark this answers is not in the document]
Comment (Trenly) [reply doc/Settings.md:274]: > The settings code could even handle both bool and object values for dependencies if we were inclined.

So you're saying it would be possible to have it where
```json
"experimentalFeatures": {
  "dependencies": true
}
```
Would either enable all dependencies, or keep the current behavior of only enabling package dependencies

OR, users could specify individual dependencies features?
```json
"experimentalFeatures": {
  "dependencies": {
    "packageDependencies" : true
    "windowsFeatures": false
  }
}
```

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 6]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+
+            std::vector<std::string> invalidFeatures;
+
+            rootDependencies.ApplyToType(DependencyType::WindowsFeature, [&invalidFeatures](Dependency dependency)
+                {
+                    const auto& name = dependency.Id();
+                    WindowsFeature::WindowsFeature windowsFeature{ name };
+                    if (!windowsFeature.DoesExist())
+                    {
+                        invalidFeatures.emplace_back(name);
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:168]: Instead of using memory to create a vector, why not have a string like: `<value>Windows Feature not found: {0}</value>` and list the features out as they are found to be invalid, and make invalidFeatures a bool; Then, later you can show the message `<value>One or more dependencies were not found; proceeding due to --ignore-missing-dependencies.</value>`  (or the equivalent for not proceeding) which could be used for any dependency type.

Then you wouldn't need the bool for whether or not to terminate, and avoid the iterator for invalidFeatures

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 7]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+                            }
+                        }
+                    }
+            });
+
+            if (FAILED(hr))
+            {
+                if (continueOnFailure)
+                {
+                    context.Reporter.Warn() << Resource::String::FailedToEnableWindowsFeatureOverridden << std::endl;
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:226]: Other than the log, there isn't any indication of *which* feature(s) failed to enable; Users may want this information to attempt manually enabling the feature without having to dig into the log.

File: src/AppInstallerCLITests/TestData/InstallFlowTest_WindowsFeatures.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      InstallerType: exe
+      InstallerSha256: 65DB2F2AC2686C7F2FD69D4A4C6683B888DC55BFA20A0E32CA9F838B51689A3B
+      InstallerSwitches:
+        Custom: /custom /scope=machine
+        SilentWithProgress: /silentwithprogress
+        Silent: /silence
+        Update: /update
+      Dependencies:
+        WindowsFeatures:
+        - netfx3
```
Comment (Trenly) [inline src/AppInstallerCLITests/TestData/InstallFlowTest_WindowsFeatures.yaml:21]: Does casing matter for windows features? If it does, it probably wouldn't be a bad idea to have a test that validates incorrect casing fails.

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 9]
```diff
@@ -133,6 +135,109 @@ namespace AppInstaller::CLI::Workflow
...
+                }
+                else
+                {
+                    context.Reporter.Error() << Resource::String::FailedToEnableWindowsFeatureOverrideRequired << std::endl;
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:230]: @denelon - I may be asking the wrong (scope creep) questions here, so forgive me, but would we want any sort of telemetry around these?
1) Which features are requested to be enabled
2) The success / fail rates of enabling features

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1697]: There is a symbolic name and a friendly (localized) name.  I don't know how easy it is to get the friendly name from the symbolic name.  Which do you expect to see (or both)?

File: src/AppInstallerCLICore/ExecutionArgs.h [hunk 10]
```diff
@@ -46,6 +46,7 @@ namespace AppInstaller::CLI::Execution
             AcceptPackageAgreements, // Accept all license agreements for packages
             Rename, // Renames the file of the executable. Only applies to the portable installerType
             NoUpgrade, // Install flow should not try to convert to upgrade flow upon finding existing installed version
+            IgnoreMissingDependencies, // Ignores missing Windows Feature dependencies.
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/ExecutionArgs.h:49]: Also, should `--force` apply here as well?

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp (hunk 9 above)
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:230]: I think we should be showing each feature as its own item and not aggregating them as well.

File: doc/Settings.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply doc/Settings.md:274]: Yes, it would be possible to write the code the enable this.  I'm not signing @ryfu-msft up for that though 😄

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp [hunk 11]
```diff
@@ -495,6 +495,7 @@ namespace AppInstaller::CLI::Workflow
             Workflow::GetDependenciesFromInstaller <<
             Workflow::ReportDependencies(Resource::String::InstallAndUpgradeCommandsReportDependencies) <<
             Workflow::ManagePackageDependencies(Resource::String::InstallAndUpgradeCommandsReportDependencies) <<
+            Workflow::EnableWindowsFeaturesDependencies <<
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/InstallFlow.cpp:498]: This whole thing doesn't seem like the right way to handle dependencies.  At least, "download" is not consistent with "and also, handle getting all dependencies squared away".  So at least maybe change the name of this task?

File: src/AppInstallerCLICore/ExecutionArgs.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCLICore/ExecutionArgs.h:49]: Since you are using `--force` for overriding a failed install, I think that equally applies to one that doesn't exist.  I would remove this option.

Along those lines, I wouldn't treat an unknown feature as a special failure checked up front, but rather as a special failure when it is being satisfied in the normal flow.

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        WindowsFeature(const std::string& name) : m_featureName{ name } {};
+
+        bool DoesExist();
+        bool IsEnabled();
+        HRESULT EnableFeature();
+        HRESULT DisableFeature();
+
+    private:
+        DismApiHelper m_dismApiHelper;
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/winget/WindowsFeature.h:156]: There is a lot of repeated loading and probing for the functions using this pattern.  It would be much better to just have this created one time.

If you want dism to be able to unload (which seems reasonable), then I would design this differently as well.  I would then create a single `DismHelper` object in the workflow and have it create `WindowsFeature` objects for you.

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        using DismGetFeatureInfoPtr = HRESULT(WINAPI*)(UINT, PCWSTR, PCWSTR, DismPackageIdentifier, DismFeatureInfo**);
+        using DismEnableFeaturePtr = HRESULT(WINAPI*)(UINT, PCWSTR, PCWSTR, DismPackageIdentifier, BOOL, PCWSTR*, UINT, BOOL, HANDLE, DISM_PROGRESS_CALLBACK, PVOID);
+        using DismDisableFeaturePtr = HRESULT(WINAPI*)(UINT, PCWSTR, PCWSTR, BOOL, HANDLE, DISM_PROGRESS_CALLBACK, PVOID);
+        using DismDeletePtr = HRESULT(WINAPI*)(VOID*);
+
+        typedef UINT DismSession;
+
+        wil::unique_hmodule m_module;
+        DismSession m_session = DISM_SESSION_DEFAULT;
+        DismFeatureInfo* m_featureInfo = nullptr;
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/winget/WindowsFeature.h:130]: This shouldn't be holding state that is feature specific.
You shouldn't be using a raw pointer if you can help it, use a smart pointer type instead.  Both `std::unique_ptr` and `wil::unique_any` can handle this.

File: src/AppInstallerCommonCore/WindowsFeature.cpp [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        if (s_IsWindowsFeatureEnabledResult_TestHook_Override)
+        {
+            return *s_IsWindowsFeatureEnabledResult_TestHook_Override;
+        }
+#endif
+        DismFeatureInfo* featureInfo = m_dismApiHelper.GetFeatureInfo(m_featureName);
+        DismPackageFeatureState featureState = featureInfo->FeatureState;
+        AICLI_LOG(Core, Verbose, << "Feature state of " << m_featureName << " is " << featureState);
+        return (featureState == DismStateInstalled || featureState == DismStateInstallPending);
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/WindowsFeature.cpp:205]: Do you have more information on why `DismStateInstallPending` should be included but `DismStatePartiallyInstalled` is not? My gut would be that only `DismStateInstalled` would be installed and everything else should require us to do something.

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 15]
```diff
@@ -24,6 +25,14 @@ namespace AppInstaller::Logging
         m_name = GetNameForPath(filePath);
         m_filePath = filePath;
 
+        // Prevent inheritance to ensure log file handle is not opened by other processes.
+        FILE* filePtr;
+        if (!fopen_s(&filePtr, m_filePath.string().c_str(), "w") && filePtr != NULL)
+        {
+            SetHandleInformation((HANDLE)_get_osfhandle(_fileno(filePtr)), HANDLE_FLAG_INHERIT, 0);
+            std::ofstream outFile(filePtr);
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:33]: There are two functions that open the stream; a helper should be used to update both.

This code is creating `outfile` and then immediately throwing it away.  You would want:
```
m_stream = std::ofstream{ filePtr };
```
And then to not call `m_stream.open`.

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 16]
```diff
@@ -24,6 +25,14 @@ namespace AppInstaller::Logging
         m_name = GetNameForPath(filePath);
         m_filePath = filePath;
 
+        // Prevent inheritance to ensure log file handle is not opened by other processes.
+        FILE* filePtr;
+        if (!fopen_s(&filePtr, m_filePath.string().c_str(), "w") && filePtr != NULL)
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:30]: This needs proper error handling.

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 17]
```diff
@@ -134,4 +124,22 @@ namespace AppInstaller::Logging
                 catch (...) {}
             }).detach();
     }
+
+    void FileLogger::OpenFileLoggerStream() 
+    {
+        // Prevent inheritance to ensure log file handle is not opened by other processes.
+        FILE* filePtr;
+        if (!fopen_s(&filePtr, m_filePath.string().c_str(), "w") && filePtr != NULL)
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:132]:
```suggestion
        errno_t fopenError = _wfopen_s(&filePtr, m_filePath.wstring().c_str(), L"w");
```

Use wide strings and capture the return value from `_wfopen_s` rather than using the global `errno`.  If you want to guard against a null pointer even when the return value indicates success, do that inside the block.

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 18]
```diff
@@ -134,4 +124,22 @@ namespace AppInstaller::Logging
...
             }).detach();
     }
+
+    void FileLogger::OpenFileLoggerStream() 
+    {
+        // Prevent inheritance to ensure log file handle is not opened by other processes.
+        FILE* filePtr;
+        if (!fopen_s(&filePtr, m_filePath.string().c_str(), "w") && filePtr != NULL)
+        {
+            if (!SetHandleInformation((HANDLE)_get_osfhandle(_fileno(filePtr)), HANDLE_FLAG_INHERIT, 0))
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:134]:
```suggestion
            THROW_IF_WIN32_BOOL_FALSE(SetHandleInformation(static_cast<HANDLE>(_get_osfhandle(_fileno(filePtr))), HANDLE_FLAG_INHERIT, 0));
```

File: src/AppInstallerCommonCore/FileLogger.cpp [hunk 19]
```diff
@@ -24,25 +24,15 @@ namespace AppInstaller::Logging
...
+        OpenFileLoggerStream();
     }
 
     FileLogger::FileLogger(const std::string_view fileNamePrefix)
     {
         m_name = "file";
         m_filePath = Runtime::GetPathTo(Runtime::PathName::DefaultLogLocation);
         m_filePath /= fileNamePrefix.data() + ('-' + Utility::GetCurrentTimeForFilename() + s_fileLoggerDefaultFileExt.data());
-
-        m_stream.open(m_filePath, s_fileLoggerDefaultOpenMode);
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/FileLogger.cpp:36]: I can't comment on that line, but `s_fileLoggerDefaultOpenMode` is no longer used with these changes.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 2 above)
Comment (ryfu-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: Changed to showing friendly name as well as the symbolic name such as `[netfx3]`

File: src/AppInstallerCLITests/TestData/InstallFlowTest_WindowsFeatures.yaml (hunk 8 above)
Comment (ryfu-msft) [reply src/AppInstallerCLITests/TestData/InstallFlowTest_WindowsFeatures.yaml]: casing doesn't matter, added tests to verify that

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp (hunk 7 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:213]: Changed to report each feature that failed to enable

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 20]
```diff
@@ -133,6 +135,104 @@ namespace AppInstaller::CLI::Workflow
...
+                    AICLI_TERMINATE_CONTEXT(hr);
+                }
+            }
+            else if (rebootRequired)
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:221]: I think this neglects the case where one feature fails to enable, and another has a reboot required. I don't know if we care about that case or not, or if that's even a possibility with the way `hr` gets handled.

If `--force` isn't present, I don't think it matters, since both the reboot and the fail would cause the context to terminate, and the termination reason doesn't make much of a difference. If `--force` is present, the user would only be informed they are bypassing the failed install, and they wouldn't be informed they are also bypassing the reboot requirement.

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 21]
```diff
@@ -133,6 +135,104 @@ namespace AppInstaller::CLI::Workflow
...
+            rootDependencies.ApplyToType(DependencyType::WindowsFeature, [&hr, &dismHelper, &info, &warn, &force, &rebootRequired](Dependency dependency)
+                {
+                    if (SUCCEEDED(hr) || force)
+                    {
+                        auto featureName = dependency.Id();
+                        DismHelper::WindowsFeature windowsFeature = dismHelper.CreateWindowsFeature(featureName);
+
+                        if (windowsFeature.DoesExist())
+                        {
+                            if (!windowsFeature.IsEnabled())
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:178]: I think there's a missing check for a pending reboot if the feature is already enabled?

Thinking of the case where a user attempts an install and gets a failure that a reboot is required. If they attempt to run the install command again without rebooting, the Windows Feature will be shown as enabled(?) and therefore won't trigger the rebootRequired, effectively negating the need to use `--force` by running the command twice.

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp (hunk 6 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp]: no longer create vector to aggregate missing dependencies, they will be reported at the end after each windows feature is processed

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp (hunk 11 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/InstallFlow.cpp]: Moved this out of "download" to the step right before `InstallPackageInstaller`.

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h (hunk 13 above)
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h]: I tried using a unique_ptr but I don't think it's possible to access the memory address of a unique_ptr which is what I need for the **FeatureInfo parameter in [DismFeatureInfo](https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/dism/dismgetfeatureinfo-function?view=windows-11). I believe this is okay though because DismDelete should handle the cleanup of this pointer.

File: src/AppInstallerCommonCore/WindowsFeature.cpp (hunk 14 above)
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/WindowsFeature.cpp]: I changed the condition so that only `DismStateInstalled` is considered as enabled.

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp (hunk 9 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:217]: Changed so that each windows feature is represented as its own item.

Comment (Trenly) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp:217]: Sounds good; I'm presuming any telemetry needs can be added once support is no longer experimental

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/WindowsFeature.cpp [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+        }
+#endif
+        return WindowsFeature(name, m_dismGetFeatureInfo, m_dismEnableFeature, m_dismDisableFeature, m_dismDelete, m_dismSession);
+    }
+
+    void DismHelper::Initialize()
+    {
+        if (m_dismInitialize)
+        {
+            LOG_IF_FAILED(m_dismInitialize(2, NULL, NULL));
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/WindowsFeature.cpp]: Changed to using enum value `DismLogErrorsWarningsInfo`

File: src/AppInstallerCommonCore/WindowsFeature.cpp [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+            AICLI_LOG(Core, Error, << "Could not load dismapi.dll");
+            return;
+        }
+
+        m_dismInitialize =
+            reinterpret_cast<DismInitializePtr>(GetProcAddress(m_module.get(), "DismInitialize"));
+        if (!m_dismInitialize)
+        {
+            AICLI_LOG(Core, Error, << "Could not get proc address of DismInitialize");
+            return;
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/WindowsFeature.cpp]: Changed to using suggested helper function. I chose to use `THROW_LAST_ERROR` since that returns the error code returned from `GetProcAddress`

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+                DismEnableFeaturePtr enableFeaturePtr,
+                DismDisableFeaturePtr disableFeaturePtr,
+                DismDeletePtr deletePtr,
+                DismSession session);
+
+            void GetFeatureInfo();
+
+            std::string m_featureName;
+            DismFeatureInfo* m_featureInfo = nullptr;
+            DismGetFeatureInfoPtr m_getFeatureInfoPtr = nullptr;
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h]: I moved WindowsFeature out of DismHelper and created a `shared_ptr<DismHelper>` member. WindowsFeature now calls the functions through the dismhelper shared pointer.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 25]
```diff
@@ -1782,4 +1782,33 @@ Please specify one of them using the --source option to proceed.</value>
...
+    <value>Failed to enable Windows Feature dependencies; proceeding due to --force</value>
+    <comment>"Windows Feature(s)" is the name of the options Windows features setting.
+{Locked="--force"}</comment>
+  </data>
+  <data name="EnablingWindowsFeature" xml:space="preserve">
+    <value>Enabling {0} [{1}]...</value>
+    <comment>{Locked="{0}","{1}"} Message displayed to the user regarding which Windows Feature is being enabled.</comment>
+  </data>
+  <data name="FailedToEnableWindowsFeature" xml:space="preserve">
+    <value>Failed to enable [{0}] feature.</value>
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: Added hresult error code to output

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 26]
```diff
@@ -1782,4 +1782,33 @@ Please specify one of them using the --source option to proceed.</value>
...
+  <data name="FailedToEnableWindowsFeatureOverrideRequired" xml:space="preserve">
+    <value>Failed to enable Windows Feature dependencies. To proceed with installation, use --force</value>
+    <comment>{Locked="--force"}</comment>
+  </data>
+  <data name="RebootRequiredToEnableWindowsFeatureOverrideRequired" xml:space="preserve">
+    <value>Reboot required to fully enable the Windows Feature(s); to override this check use --force</value>
+    <comment>"Windows Feature(s)" is the name of the options Windows features setting.</comment>
+  </data>
+  <data name="FailedToEnableWindowsFeatureOverridden" xml:space="preserve">
+    <value>Failed to enable Windows Feature dependencies; proceeding due to --force</value>
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1798]: I left this as is because there are many other examples where `proceeding due to --arg` is used such as:
InstallerHashMismatchOverridden
PortableHashMismatchOverridden
ArchiveFailedMalwareScanOverridden

I think this is fine because it tells the user what arg they included to override the check. I'm also not sure what else to put so if you feel strongly about this, I can change this (and the others as well) to follow a different pattern.

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.h [hunk 27]
```diff
@@ -59,4 +59,6 @@ namespace AppInstaller::CLI::Workflow
     // Inputs: PackageVersion, Manifest
     // Outputs: DependencySource
     void OpenDependencySource(Execution::Context& context);
+
+    void EnableWindowsFeaturesDependencies(Execution::Context& context);
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.h:67]: Added workflow comments

File: src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp [hunk 28]
```diff
@@ -133,6 +135,104 @@ namespace AppInstaller::CLI::Workflow
...
+                        {
+                            // Note: If a feature is not found, continue enabling the rest of the dependencies but block immediately after unless force arg is present.
+                            AICLI_LOG(Core, Info, << "Windows Feature [" << featureName << "] does not exist");
+                            hr = APPINSTALLER_CLI_ERROR_INSTALL_MISSING_DEPENDENCY;
+                            warn << Resource::String::WindowsFeatureNotFound(Utility::LocIndView{ featureName }) << std::endl;
+                        }
+                    }
+            });
+
+            if (FAILED(hr) || hr == APPINSTALLER_CLI_ERROR_INSTALL_MISSING_DEPENDENCY)
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp]: Removed APPINSTALLER_CLI_ERROR_INSTALL_MISSING_DEPENDENCY  since you are right and it is covered by `FAILED(hr)`

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+        /// <summary>
+        /// Struct representation of a single Windows Feature.
+        /// </summary>
+        struct WindowsFeature
+        {
+            HRESULT Enable();
+            HRESULT Disable();
+            bool DoesExist();
+            bool IsEnabled();
+            std::wstring GetDisplayName();
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h]: Changed to returning a LocIndString directly.

File: src/AppInstallerCLICore/Workflows/InstallFlow.cpp (hunk 11 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/InstallFlow.cpp:498]: Thanks for the explanation, I didn't know that COM allowed for multiple downloads. I moved the task back to the suggested location and added a TODO comment explaining the issue that needs to be addressed.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 30]
```diff
@@ -1804,8 +1804,9 @@ Please specify one of them using the --source option to proceed.</value>
     <comment>{Locked="{0}","{1}"} Message displayed to the user regarding which Windows Feature is being enabled.</comment>
   </data>
   <data name="FailedToEnableWindowsFeature" xml:space="preserve">
-    <value>Failed to enable [{0}] feature.</value>
-    <comment>{Locked="{0}"} Error message displayed when a failure was encountered when enabling a Windows Feature.</comment>
+    <value>Failed to enable [{0}] feature with exit code: {1}</value>
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: Changed to suggested, I used GetPresentableUserMessage to handle `{2}: {3}`

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 31]
```diff
@@ -113,69 +115,17 @@ namespace AppInstaller::WindowsFeature
...
-        {
-            CloseSession();
-            Shutdown();
-        };
-
-        WindowsFeature CreateWindowsFeature(const std::string& name);
+        ~DismHelper();
 
     private:
+        friend WindowsFeature;
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h]: removed and exposed public functions

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h [hunk 32]
```diff
@@ -194,5 +144,46 @@ namespace AppInstaller::WindowsFeature
...
+    };
+
+    /// <summary>
+    /// Struct representation of a single Windows Feature.
+    /// </summary>
+    struct WindowsFeature
+    {
+        // This default constructor is only used for mocking unit tests.
+        WindowsFeature() = default;
+        WindowsFeature(std::shared_ptr<DismHelper> dismHelper, const std::string& name);
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h:178]: I left the constructor as is to be able to get a shared_ptr to DismHelper. I changed DismHelper to derive from `enable_shared_from_this` so that we can safely get another shared_ptr as suggested.

[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h:178]: The awkward part is the call site; the need to pass the `shared_ptr` yourself.  (I should have put the comment there I suppose...)  I wasn't suggesting that you should take the constructor away (although you could make it protected and friend the other way).  I *was* suggesting that you add a function like:
```
WindowsFeature DismHelper::GetFeature(std::string featureName);
```
So that the call site was less awkward, like:
```
WindowsFeature::WindowsFeature windowsFeature = dismHelper->GetFeature(featureName};
```
If you did make the constructor protected, you then don't have to worry about the `shared_ptr` being null.

**Review by JohnMcPMS — APPROVED** [review]

File: src/AppInstallerCommonCore/Public/winget/WindowsFeature.h (hunk 32 above)
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Public/winget/WindowsFeature.h]: I agree the call site is weird passing shared ptrs around. I feel that this is important if dismhelper were to be extended to other scenarios so I changed it based on your suggestions. I created a helper function within dismhelper that creates the windows object for you, calling the constructor with the shared ptr. I made the WindowsFeature constructor protected and made dismHelper a friend so that only the dismhelper can create the WindowsFeature object.
